### PR TITLE
fix(ce-code-review): tighten autofix_class rubric for safe_auto/gated_auto boundary

### LINKED ISSUE #686: Calibrate safe_auto vs gated_auto boundary in ce-code-review persona output
## Context

PR #685 (`fix(ce-code-review): drop LFG defer-bias and approval gates`) shipped a fix for the LFG path's defer-bias: personas now propose `suggested_fix` more aggressively, and the LFG path applies findings immediately instead of routing to ticket-filing. That work also surfaced a separate concern about headless mode and the `autofix_class` taxonomy that was deliberately deferred from PR #685.

This issue tracks that follow-up.

## The concern

Headless mode (used by `ce-work` for inline quality checks during execution, and other programmatic callers) only auto-applies findings classified as `safe_auto`. Everything else — `gated_auto`, `manual`, `advisory` — is returned in the structured envelope for the caller to route.

The `safe_auto` definition is "local and deterministic — the fixer can apply it mechanically without design judgment." Examples in the subagent template: extracting a duplicated helper, adding a missing nil/null check, fixing an off-by-one, adding a missing test for an untested code path, removing dead code.

PR #685 did **not** touch the `safe_auto` vs `gated_auto` boundary. It only pushed personas to populate `suggested_fix` more often. So findings that were previously classified as `manual` without a fix may now be `manual` with a fix, or `gated_auto` with a fix — but very few, if any, additional findings flow into `safe_auto`.

The hypothesis: personas are conservative about `safe_auto` because the wrong-side cost feels asymmetric — auto-applying something that shouldn't have been is more visible than failing to auto-apply something that could have been. This pushes real `safe_auto` candidates into `gated_auto` or `manual`, which means:

- `ce-work`'s inline auto-quality-check (using headless) is weaker than it could be — fewer fixes land automatically during work, more end up flagged for the user later.
- Headless callers that consume the envelope have to do more routing themselves.

## Why this matters

The original incident behind PR #685 ("8 findings ended up in tickets that the agent could fix") was about LFG specifically. But a related shape can recur in autonomous flows: ce-work runs headless, gets back a `manual + suggested_fix` finding that's actually mechanical and safe, and routes it as residual instead of just applying it. The user then runs interactive review later and sees a finding that should have been gone.

## Why we deferred this from PR #685

Three reasons:

1. **Scope.** PR #685 was already a meaningful change (LFG path + persona-side push + 6 implementation units). Adding a `safe_auto` rubric tightening would have made it harder to evaluate either piece independently.
2. **Empirical data.** PR #685 makes the classification visible per finding (every finding now has `suggested_fix` populated more often). Real review runs against the new infrastructure will show whether `safe_auto` is genuinely undersized or whether the current rubric is right.
3. **Risk.** Tightening `safe_auto` is a higher-risk change than the LFG path edits — it affects autofix mode and headless callers in production-adjacent flows. Worth doing carefully with data, not as a rider on an unrelated change.

## What an agent picking this up should do

The high-level shape:

1. **Gather data.** Run `/ce-code-review` (interactive mode, on a few real branches) after PR #685 has merged, capture the per-agent artifact files in `.context/compound-engineering/ce-code-review/<run-id>/`, and look at the distribution of findings across `safe_auto` / `gated_auto` / `manual` / `advisory`. Specifically: for each `gated_auto` and `manual + suggested_fix` finding, ask whether it could have been `safe_auto` — meaning the fix is mechanical (no design judgment), purely additive or matches a parallel pattern in the codebase, and doesn't change contracts/permissions/module boundaries.
2. **Identify under-classifications.** Build a list of finding types that personas are routing to `gated_auto` or `manual` that the rubric would actually permit at `safe_auto`. Common candidates to look for:
   - Adding a missing null check inside a function (when no public-API contract changes)
   - Fixing an obvious off-by-one when the parallel pattern is in scope
   - Adding a missing test for an untested code path
   - Removing dead code
   - Extracting a duplicated helper when the duplication is local
3. **Tighten the rubric.** Update `plugins/compound-engineering/skills/ce-code-review/references/subagent-template.md` (lines ~137-142, the `autofix_class` decision guide) and the schema description in `references/findings-schema.json`. Provide concrete examples of the boundary cases you found in step 2 — show personas the kinds of findings that *should* be `safe_auto` even when they feel risky to commit to.
4. **Test against the prior runs.** Re-run the review on the same branches you collected data from in step 1; confirm the distribution shifts toward `safe_auto` for the cases you targeted, and confirm no `gated_auto` / `manual` cases that genuinely deserve approval get pulled down by the tightened rubric.
5. **Propose the change.** Write up the calibration findings as a `docs/solutions/skill-design/` entry (per `AGENTS.md` repo conventions for capturing design learnings). Open a PR with the rubric change.

## Constraints and gotchas

- **Don't change headless's behavior contract.** This issue is about *which findings end up in `safe_auto`*, not about what headless does with them. Headless still auto-applies only `safe_auto`. `ce-work` and other callers depend on that.
- **Don't change the LFG path.** PR #685 is the source of truth for LFG behavior. Don't reintroduce gates or change the heterogeneous fixer queue contract.
- **Persona changes propagate to all modes.** Tightening the persona prompt affects autofix, headless, interactive, walk-through — every path. The whole point is uniform classification, but it does mean small persona-prompt edits have broad blast radius.
- **`docs/plans/2026-04-25-001-fix-ce-code-review-lfg-defer-bias-plan.md`** is the plan PR #685 implemented. Read it for full context on persona-side push and the `suggested_fix` rule.
- **`docs/plans/2026-04-21-002-refactor-ce-code-review-precision-and-validation-plan.md`** is the prior plan that introduced the anchored confidence rubric and Stage 5b. Read for context on how `safe_auto` vs `gated_auto` boundary was last calibrated.
- **`docs/solutions/skill-design/confidence-anchored-scoring-2026-04-21.md`** documents the anchored-confidence-score reasoning. Calibration writeups follow this style.

## Files likely to change

- `plugins/compound-engineering/skills/ce-code-review/references/subagent-template.md` — `autofix_class` decision guide, lines around 137-142
- `plugins/compound-engineering/skills/ce-code-review/references/findings-schema.json` — `autofix_class` description, possibly add examples to `_meta`
- New: `docs/solutions/skill-design/<date>-safe-auto-rubric-calibration.md` — calibration writeup
- `tests/review-skill-contract.test.ts` — new assertions for the tightened rubric

## Definition of done

- Persona prompt and schema updated with concrete `safe_auto` examples drawn from real review runs.
- A calibration writeup in `docs/solutions/skill-design/` documenting the data and reasoning.
- Tests assert the tightened rubric is in place.
- Re-run reviews on a few branches show `safe_auto` capturing genuinely-mechanical findings that previously routed to `gated_auto` / `manual`.

## Related

- PR #685 — `fix(ce-code-review): drop LFG defer-bias and approval gates`
- Plan: `docs/plans/2026-04-25-001-fix-ce-code-review-lfg-defer-bias-plan.md`
- Prior plan: `docs/plans/2026-04-21-002-refactor-ce-code-review-precision-and-validation-plan.md`

## Notes for picking up in a future session

This issue body is self-contained — a fresh agent session should be able to read it and start work without prior conversation context. The repo's `AGENTS.md` and `plugins/compound-engineering/AGENTS.md` carry the contributor conventions (commit format, no manual version bumps, branching policy, etc.). Use `/ce-plan` or `/ce-brainstorm` if the calibration approach needs more structure than this issue provides.

### PR DESCRIPTION
**Pull request by tmchow**
Closes #686.

## TL;DR

Issue #686 hypothesized that personas under-classify findings as `safe_auto`, making `ce-work`'s headless auto-apply weaker than it could be. **The hypothesis turned out to be approximately wrong** — a 60-trial synthetic-fixture eval shows the post-#685 baseline already classifies textbook mechanical cases as `safe_auto`. 6 of 9 fixture shapes are identical across baseline and tightened rubric.

The actual win the eval surfaced is **variance reduction on ambiguous cases**, not a `safe_auto`-rate increase. Headline: orphan code without explicit "no callers" annotation. Baseline produced 3 different classifications across 4 trials on the same input (essentially random); tightened pins it deterministically to `gated_auto`.

## What changed

| File | Lines | Description |
|---|---|---|
| `subagent-template.md` | +14 / −6 | Decision-guide expansion: one-sentence symmetry-of-error framing, operational test for `safe_auto`, four "boundary cases that often feel risky but are still safe_auto" examples, anti-default guard for `gated_auto` parallel to existing one for `advisory`. |
| `findings-schema.json` | +1 / −1 | `autofix_class` field description rewritten to mirror the operational test from the subagent template. |
| `tests/review-skill-contract.test.ts` | +28 / 0 | New test asserting the rubric's new structural elements (boundary cases, operational test phrasing, anti-default guards). |
| `docs/solutions/skill-design/safe-auto-rubric-calibration-2026-04-25.md` | +127 / 0 | Calibration writeup: eval methodology, 60-trial results matrix, the methodological lesson (calibrations should be evaluated for variance reduction first, rate-shift second), reproducibility notes. |

## Eval results — 60 trials, 9 fixture shapes

| Fixture | Baseline | Tightened | Verdict |
|---|---|---|---|
| F1 — internal nil guard | 3/3 safe_auto | 3/3 safe_auto | identical |
| F1b — cart `min_by` semantic bug | 3/3 safe_auto | 3/3 safe_auto | identical |
| F2 — off-by-one + parallel pattern | 3/3 safe_auto | 3/3 safe_auto | identical |
| F3 — dead code w/ "no callers" comment | 3/3 safe_auto | 3/3 safe_auto | identical |
| F4 — local helper extract | 2/3 safe_auto, 1/3 advisory | 3/3 safe_auto | **tightened reduces variance** |
| **F3b** — orphan code, no explicit comment | manual / safe_auto / gated_auto / safe_auto (4 trials, 3 classes) | 7/7 gated_auto | **tightened dramatically reduces variance** |
| F4b — cross-file Rails service extract | 4/4 safe_auto | 6/7 gated_auto, 1/7 advisory | stable disagreement, both defensible |
| F5 — missing test for new public method | 3/3 safe_auto | 3/3 safe_auto | identical |
| F6 — admin auth gate (negative control) | 1/1 gated_auto | 1/1 gated_auto | identical (correctly stable) |

## The F4b trade-off

Cross-file extraction of two service objects with identical bodies. Baseline picks `safe_auto` (matches "extracting a duplicated helper" example). Tightened picks `gated_auto` (matches "naming/placement requires a design conversation" — Rails service-layering placement is a real architectural call). Both are internally consistent.

Tightened picks the conservative reading, meaning ce-work's headless will flag cross-file extraction for user review instead of auto-applying it. For careful operators that's the right call; for autonomous bulk-refactor flows it's modestly more friction. Documented as a known trade-off in the calibration writeup.

## What this PR doesn't do

The eval is single-persona on synthetic fixtures. Real reviews run multiple personas through synthesis with conservative tie-breaks; synthesis-layer effects could amplify or dampen what the persona-side eval shows. If a `safe_auto` underclassification incident recurs on a real branch (the original "8 findings to tickets" story), that's evidence for another iteration.

## Test plan

- [x] `bun test` — 910/910 pass (1 new contract test, 12 new assertions)
- [x] `bun run release:validate` — clean
- [x] Confirmed: the new rubric language is referenced in tests so future drift fails the contract test

## References

- #685 — `fix(ce-code-review): replace LFG with best-judgment auto-resolve` (the suggested_fix push this builds on)
- `docs/solutions/skill-design/safe-auto-rubric-calibration-2026-04-25.md` — full calibration writeup
- `docs/solutions/skill-design/confidence-anchored-scoring-2026-04-21.md` — the anchored confidence rubric this shares stylistic conventions with

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: docs/solutions/skill-design/safe-auto-rubric-calibration-2026-04-25.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+tags:
+  - ce-code-review
+  - autofix-class
+  - rubric
+  - calibration
+  - eval
+  - eval-methodology
+  - variance
+related_issue: EveryInc/compound-engineering-plugin#686
+related_pr: PR #685 (suggested_fix push that this builds on)
```
[the remark this answers is not in the document]
Comment (tmchow) [reply docs/solutions/skill-design/safe-auto-rubric-calibration-2026-04-25.md]: > The `related_pr` frontmatter currently uses `PR #685 ...` unquoted, and in YAML a space before `#` starts a comment, so parsers will read this field as just `PR` and silently drop the PR number/context.

Addressed in 5fdf01f5: quoted the `related_pr` value so it parses as the full string. While in here I also swept `docs/` for the same pattern and found one additional offender (`docs/plans/2026-04-16-001-fix-ce-polish-beta-detection-gaps-plan.md`'s `title:` field), which I quoted too. Verified both files parse correctly via `yaml.safe_load`; repo-wide re-sweep shows zero remaining ' #' instances in frontmatter values.
